lib/oidc/service.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require 'common/client/concerns/monitoring'
+require 'oidc/configuration'
+require 'oidc/response'
+module OIDC
+  class Service < Common::Client::Base
+    include Common::Client::Concerns::Monitoring
+    configuration OIDC::Configuration
+
+    STATSD_KEY_PREFIX = 'oidc'
+
+    def call_no_token(action, url)
+      connection.send(action) do |req|
+        req.url url
+        req.headers['Content-Type'] = 'application/json'
+        req.headers['Accept'] = 'application/json'
+      end
+    end
+
+    def metadata(iss)
+      metadata_endpoint = get_metadata_endpoint(iss)
+      with_monitoring do
+        OIDC::Response.new call_no_token('get', metadata_endpoint)
+      end
+    end
+
+    def oidc_jwks_keys(iss)
+      url = metadata(iss).body['jwks_uri']
+      with_monitoring do
+        call_no_token('get', url)
+      end
+    end
+
+    def get_metadata_endpoint(iss)
+      metadata_endpoint = Settings.oidc.issuers.find { |s| iss.downcase.include? s['prefix'].downcase }
```

Comment:
Two observations...

I think this should be `starts_with` instead of `include`, and I think it needs some error handling if it doesn't know how to proxy the request.

Given:
```
irb(main):042:0> issuers = [ { 'prefix' => 'https://www.example.com/known-issuer', 'proxy' => 'https://www.example.com/proxy' } ]
=> [{"prefix"=>"https://www.example.com/known-issuer", "proxy"=>"https://www.example.com/proxy"}]
```

Happy path - request is modified and sent through the proxy
```
irb(main):049:0> iss = 'https://www.example.com/known-issuer'
=> "https://www.example.com/known-issuer"
irb(main):050:0> metadata_endpoint = issuers.find { |s| iss.downcase.include? s['prefix'].downcase }
=> {"prefix"=>"https://www.example.com/known-issuer", "proxy"=>"https://www.example.com/proxy"}
irb(main):051:0> proxied_iss = iss.gsub(metadata_endpoint['prefix'], metadata_endpoint['proxy'])
=> "https://www.example.com/proxy"
```

Known issuer, but in random location - request is modified but does not go through the proxy
```
irb(main):056:0> iss = 'https://www.example.com/unknown-issuer?q=https://www.example.com/known-issuer'
=> "https://www.example.com/unknown-issuer?q=https://www.example.com/known-issuer"
irb(main):057:0> metadata_endpoint = issuers.find { |s| iss.downcase.include? s['prefix'].downcase }
=> {"prefix"=>"https://www.example.com/known-issuer", "proxy"=>"https://www.example.com/proxy"}
irb(main):058:0> proxied_iss = iss.gsub(metadata_endpoint['prefix'], metadata_endpoint['proxy'])
=> "https://www.example.com/unknown-issuer?q=https://www.example.com/proxy"
```

Unknown issuer - exception
```
irb(main):062:0> iss = 'https://www.example.com/unknown-issuer'
=> "https://www.example.com/unknown-issuer"
irb(main):063:0> metadata_endpoint = issuers.find { |s| iss.downcase.include? s['prefix'].downcase }
=> nil
irb(main):064:0> proxied_iss = iss.gsub(metadata_endpoint['prefix'], metadata_endpoint['proxy'])
Traceback (most recent call last):
        4: from /usr/bin/irb:23:in `<main>'
        3: from /usr/bin/irb:23:in `load'
        2: from /Library/Ruby/Gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        1: from (irb):64
NoMethodError (undefined method `[]' for nil:NilClass)
```

Reply:
> and I think it needs some error handling if it doesn't know how to proxy the request.

Or maybe these do just fall through and skip the proxy? I wonder if Derek has an opinion.

Reply:
I've thought the same about requests that don't appear to go through the proxy.  There are a number of url's listed in the config yaml's that skip the fwd proxy, so I think the request would fall through and skip the proxy.

If a prefix is found, then the `iss.gsub(metadata_endpoint['prefix'], metadata_endpoint['proxy'])` should succeed.

Maybe `issuers.find` needs to return an error immediately if a match isn't found rather than relying on the NoMethodError for cases without matches?

Reply:
The `app/models/token.rb` `valid_issuer?` method will need to be updated to check the different possible issuers as well.
https://github.com/department-of-veterans-affairs/vets-api/blob/master/app/models/token.rb#L50

With that check in place, the `get_metadata_endpoint` should be able to focus on just translating the issuer to the fwd proxied url.

Reply:
I see your points. As Derek mentioned if when we update the `valid_issuer?` method then the invalid prefix error would return
```json
{
  "errors": [
    {
      "title": "Not authorized",
      "detail": "Invalid issuer",
      "code": "401",
      "status": "401"
    }
  ]
}
```
Which will be an acceptable error, but the unknown issuer on a valid prefix is a bit more complicated. 

If we let the error bubble up we would receive

```json
{
  "errors": [
    {
      "title": "Internal server error",
      "detail": "Internal server error",
      "code": "500",
      "status": "500",
      "meta": {
        "exception": "undefined method `index' for nil:NilClass",
        "back_trace": [..]
      }
    }
  ]
}  
```
I'm unsure if this is ok? We could throw an invalid issuer error?

TODO
- [ ] issuer check should be `starts_with?` 
- [ ] add error handling for invalid prefix
- [ ] add error handling for unkown_issuer

Reply:
Re-grouping my thoughts on this...

This metadata-fetching-functionality is generic (supports any issuer) but that issuer must be whitelisted in the config, otherwise it has no idea which endpoint (`openid-configuration`, `oauth-authorization-server`, etc) to build/query.

Based on that, I think the question about fall-through is not applicable.

If a matching config entry can not be found, an error should be gracefully raised. The caller is responsible for deciding how they want to handle it.

If a matching config entry is found, it should be expected to have a proxy route and a metadata endpoint.